src/Compilers/CSharp/Portable/Binder/Semantics/Operators/BinaryOperatorOverloadResolution.cs

```diff
@@ -60,24 +59,62 @@ public void BinaryOperatorOverloadResolution(BinaryOperatorKind kind, BoundExpre
             //   provided by the type of x (if any) and the candidate operators provided by the type of y (if any), 
             //   each determined using the rules of 7.3.5. Candidate operators only occur in the combined set once.
 
-            bool leftHadUserDefinedCandidate = false;
-            bool rightHadUserDefinedCandidate = false;
+            // From https://github.com/dotnet/csharplang/blob/master/meetings/2017/LDM-2017-06-27.md:
+            // - We only even look for operator implementations in interfaces if one of the operands has a type that is
+            // an interface or a type parameter with a non-empty effective base interface list.
+            // - We should look at operators from classes first, in order to avoid breaking changes.
+            // Only if there are no applicable user-defined operators from classes will we look in interfaces.
+            // If there aren't any there either, we go to built-ins.
+            // - If we find an applicable candidate in an interface, that candidate shadows all applicable operators in
+            // base interfaces: we stop looking.
+
+            bool hadUserDefinedCandidate = false;
 
             // In order to preserve backward compatibility, at first we ignore interface sources.
 
             if (!leftSourceIsInterface)
             {
-                leftHadUserDefinedCandidate = GetUserDefinedOperators(kind, leftOperatorSourceOpt, left, right, result.Results, ref useSiteDiagnostics);
+                hadUserDefinedCandidate = GetUserDefinedOperators(kind, leftOperatorSourceOpt, left, right, result.Results, ref useSiteDiagnostics);
             }
 
             if (!rightSourceIsInterface && rightOperatorSourceOpt != leftOperatorSourceOpt)
             {
                 var rightOperators = ArrayBuilder<BinaryOperatorAnalysisResult>.GetInstance();
-                rightHadUserDefinedCandidate = GetUserDefinedOperators(kind, rightOperatorSourceOpt, left, right, rightOperators, ref useSiteDiagnostics);
+                hadUserDefinedCandidate |= GetUserDefinedOperators(kind, rightOperatorSourceOpt, left, right, rightOperators, ref useSiteDiagnostics);
                 AddDistinctOperators(result.Results, rightOperators);
                 rightOperators.Free();
             }
 
+            Debug.Assert(result.Results.Count == 0 || hadUserDefinedCandidate);
+
+            // If there are no applicable candidates in classes / stuctures, try with interface sources.
+            // From https://github.com/dotnet/csharplang/blob/master/meetings/2017/LDM-2017-06-27.md:
+            // Let's not allow == and != and see what feedback we get. We'll see from prototype use whether
+            // this is an intolerable limitation, but we believe it is a decent place to land.
+            if (!hadUserDefinedCandidate &&
+                kind != BinaryOperatorKind.Equal && kind != BinaryOperatorKind.NotEqual)
+            {
+                result.Results.Clear();
```

Comment:
> If so, could useSiteDiagnostics also be non-empty?

Use-site diagnostics shouldn't be discarded, if that is what you are getting to.